executor/aggfuncs/func_group_concat.go

```diff
@@ -491,32 +517,49 @@ func (e *groupConcatDistinctOrder) UpdatePartialResult(sctx sessionctx.Context,
 			continue
 		}
 		p.valSet.Insert(joinedVal)
+		memDelta += int64(len(joinedVal))
 		sortRow := sortRow{
 			buffer:  buffer,
 			byItems: make([]*types.Datum, 0, len(e.byItems)),
 		}
 		for _, byItem := range e.byItems {
 			d, err := byItem.Expr.Eval(row)
 			if err != nil {
-				return 0, err
+				return memDelta, err
 			}
+			memDelta += GetDatumMemSize(&d)
 			sortRow.byItems = append(sortRow.byItems, d.Clone())
 		}
+		oldMem := p.topN.currSize
 		truncated := p.topN.tryToAdd(sortRow)
+		newMem := p.topN.currSize
+		memDelta += int64(newMem - oldMem)
 		if p.topN.err != nil {
-			return 0, p.topN.err
+			return memDelta, p.topN.err
 		}
 		if truncated {
 			if err := e.handleTruncateError(sctx); err != nil {
-				return 0, err
+				return memDelta, err
 			}
 		}
 	}
-	return 0, nil
+	return memDelta, nil
 }
 
 func (e *groupConcatDistinctOrder) MergePartialResult(sctx sessionctx.Context, src, dst PartialResult) (memDelta int64, err error) {
 	// If order by exists, the parallel hash aggregation is forbidden in executorBuilder.buildHashAgg.
 	// So MergePartialResult will not be called.
 	return 0, terror.ClassOptimizer.New(mysql.ErrInternal, mysql.MySQLErrName[mysql.ErrInternal]).GenWithStack("groupConcatDistinctOrder.MergePartialResult should not be called")
 }
+
+func GetDatumMemSize(d *types.Datum) int64 {
+	var datumMemSize int64
+	datumMemSize += int64(unsafe.Sizeof(d.Kind()))
```

Comment:
the `Datum.Clone` operation will copy all the fields to the new one, should we need to calculate all of those? 
I think we can use 
```
datumMemSize += int64(unsafe.Sizeof(*d))
datumMemSize += int64(len(d.Collation()))
datumMemSize += int64(len(d.GetBytes()))
datumMemSize += getValMemDelta(d.GetValue()) - DefInterfaceSize
```

What your opinions? @b41sh

Reply:
If the value of **datum.k** is **KindString**, the memory usage of **datum value** will be calculated twice? 
`datumMemSize += int64(len(d.GetBytes()))`
`datumMemSize += getValMemDelta(d.GetValue()) - DefInterfaceSize`

Reply:
I'm wrong. I think we should use follow code
`datumMemSize += getValMemDelta(d.GetValue()) - DefInterfaceSize` -> `datumMemSize += getValMemDelta(d.GetInterface()) - DefInterfaceSize`
It will just calculating the memory usage of the interface. What is your opinion?

Reply:
If the value of **datum.k** is **KindInt64**, **datum value** will assign to **datum.i**. 
`getValMemDelta(d.GetInterface()) - DefInterfaceSize`
This code return 0,  **datum value** will not be  added to **datumMemSize**. 

 Follow code is ok?
`datumMemSize += int64(unsafe.Sizeof(*d))`
`datumMemSize += int64(len(d.Collation()))`
`datumMemSize += getValMemDelta(d.GetValue()) - DefInterfaceSize`

Reply:
`int64(unsafe.Sizeof(*d))` will return all memory size of the `datum` including `datum.i`. So `getValMemDelta(d.GetValue()) - DefInterfaceSize` get 0 is a correct result.

Reply:
`int64(unsafe.Sizeof(*d))` return the variable type's memory size in `Datum`,  it always be 72 bytes. In my opinion,
`datumMemSize = variable type's memory size + variable value's memory size`

However, `getValMemDelta(d.GetInterface()) - DefInterfaceSize` just return the memory size `Datum.x`'s value ,  not include memory size of other variable value in `Datum`. Is there something wrong with my understanding?

Reply:
I'm agree your opinion.
should we use follow code?
```
datumMemSize += int64(unsafe.Sizeof(*d))
datumMemSize += int64(len(d.Collation()))
datumMemSize += int64(len(d.GetBytes()))
datumMemSize += getValMemDelta(d.GetInterface()) - DefInterfaceSize
```

Reply:
PTAL @b41sh